### PR TITLE
Implement full caching

### LINKED ISSUE #45: Slow
Even opening moderately small files can cause up to a second lag. 

Here's the timings for opening the file "seaborn/categorical.py" from the popular Python plotting package Seaborn. Per the documentation (:help profile), the "self" time will be wrong since SimpylFold is called recursively.

```
FUNCTION  SimpylFold()
Called 6982 times
Total time:   0.809100
Self time:   0.360778

count  total (s)   self (s)

                                " If we are starting a new sweep of the buffer (i.e. the current line
                                " being folded comes before the previous line that was folded), initialize
                                " the cache of results of calls to `s:NumContainingDefs`
6982              0.016330     if !exists('b:last_folded_line') || b:last_folded_line > a:lnum
50              0.000235         let b:cache_NumContainingDefs = {}
50              0.000052         let b:in_docstring = 0
50              0.000013     endif
6982              0.008365     let b:last_folded_line = a:lnum

                                " If this line is blank, its fold level is equal to the minimum of its
                                " neighbors' fold levels, but if the next line begins a definition, then
                                " this line should fold at one level below the next
6982              0.010268     let line = getline(a:lnum)
6982              0.024489     if line =~ s:blank_regex
1944              0.003536         let next_line = nextnonblank(a:lnum)
1944              0.001479         if next_line == 0
                                        return 0
                                    elseif getline(next_line) =~# s:def_regex
168              0.000292             return SimpylFold(next_line) - 1
                                    else
1776              0.001048             return -1
                                    endif
                                endif

5038              0.013018     let fold_docstrings = !exists('g:SimpylFold_fold_docstring') || g:SimpylFold_fold_docstring
5038              0.031019     let docstring_match = matchlist(line, s:docstring_start_regex)
5038              0.009477     let prev_line = getline(a:lnum - 1)
5038              0.033692     if !b:in_docstring && ( prev_line =~# s:def_regex || prev_line =~ s:multiline_def_end_regex ) && len(docstring_match)
    4   0.000179   0.000013         let this_fl = s:NumContainingDefs(a:lnum) + fold_docstrings
    4              0.000004         let b:in_docstring = 1
    4              0.000006         if docstring_match[1] == '"""'
    4              0.000008             let b:docstring_end_regex = s:docstring_end_double_regex
    4              0.000001         else
                                        let b:docstring_end_regex = s:docstring_end_single_regex
                                    endif
    4              0.000003     elseif b:in_docstring
10   0.000444   0.000036         let this_fl = s:NumContainingDefs(a:lnum) + fold_docstrings
10              0.000036         if line =~ b:docstring_end_regex
    4              0.000006             let b:in_docstring = 0
    4              0.000002         endif
10              0.000005     else
                                    " Otherwise, its fold level is equal to its number of containing
                                    " definitions, plus 1, if this line starts a definition of its own
5024   0.385522   0.035029         let this_fl = s:NumContainingDefs(a:lnum) + (line =~# s:def_regex)

5024              0.002473     endif
                                " If the very next line starts a definition with the same fold level as
                                " this one, explicitly indicate that a fold ends here
5038              0.024561     if getline(a:lnum + 1) =~# s:def_regex && SimpylFold(a:lnum + 1) == this_fl
                                    return '<' . this_fl
                                else
5038              0.003546         return this_fl
                                endif


FUNCTION  <SNR>103_NumContainingDefs()
Called 6380 times
Total time:   0.356571
Self time:   0.351067

count  total (s)   self (s)

                                " Recall memoized result if it exists in the cache
6380              0.011093     if has_key(b:cache_NumContainingDefs, a:lnum)
1382              0.001994         return b:cache_NumContainingDefs[a:lnum]
                                endif

4998              0.007690     let this_ind = indent(a:lnum)

4998              0.003350     if this_ind == 0
250              0.000129         return 0
                                endif

                                " Walk backwards to the previous non-blank line with a lower indent level
                                " than this line
4748              0.005298     let i = a:lnum - 1
15716              0.006771     while 1
15716              0.060302         if getline(i) !~ s:blank_regex
11524              0.014986             let i_ind = indent(i)
11524              0.008927             if i_ind < this_ind
1342              0.002669                 let ncd = s:NumContainingDefs(i) + (getline(i) =~# s:def_regex)
1342              0.000836                 break
                                        elseif i_ind == this_ind && has_key(b:cache_NumContainingDefs, i)
3406              0.005980                 let ncd = b:cache_NumContainingDefs[i]
3406              0.001698                 break
                                        endif
6776              0.002618         endif

10968              0.008147         let i -= 1

                                    " If we hit the beginning of the buffer before finding a line with a
                                    " lower indent level, there must be no definitions containing this
                                    " line. This explicit check is required to prevent infinite looping in
                                    " the syntactically invalid pathological case in which the first line
                                    " or lines has an indent level greater than 0.
10968              0.005700         if i <= 1
                                        let ncd = getline(1) =~# s:def_regex
                                        break
                                    endif

10968              0.004633     endwhile

                                " Memoize the return value to avoid duplication of effort on subsequent
                                " lines
4748              0.010655     let b:cache_NumContainingDefs[a:lnum] = ncd

4748              0.003551     return ncd
```

### PR DESCRIPTION
**Pull request by nfnty**
Improves performance by several magnitudes.

Fixes #45

Problems:
- Breaks previous hack whereby certain lines would be extra indented even though breaking PEP8. Can possibly be fixed, but would make performance worse.
- Not currently auto recaching. Previous solution causes slowdowns and multiple recaches in a row. I've been experimenting with `autocmd`, but haven't been able to fix it properly. Not sure what the best way to fix this is currently.

### CODE REVIEW

**Comment by nfnty** [comment]
Auto recaching has been solved.

**Comment by nfnty** [comment]
I've now completely profiled the script.

**Comment by nfnty** [comment]
Fixed the multiline def extra indent hack.

**Comment by tmhedberg** [comment]
Thanks again for your work on this.

I have not yet reviewed the PR thoroughly, but I've run into an initial problem. When I use [this file](https://raw.githubusercontent.com/mwaskom/seaborn/master/seaborn/categorical.py), which is mentioned in #45 as a file exhibiting load-time slowness, to test your changes, nothing is folded. `foldmethod`, `foldexpr`, etc. appear to be set correctly, but there are no folds defined in the file. If I check out my own master branch, without your changes, the file does load with a short delay, but the folding works.

Are you able to reproduce this problem?

**Comment by nfnty** [comment]
Strange, I can't reproduce your issue. The file folds just fine.

**Comment by nfnty** [comment]
Are you only having issues with that file?

**Comment by nfnty** [comment]
Add this to your `.vimrc` or `init.vim` before the plugin is being sourced (replace `$PATH_TO_PLUGIN`):
```
profile start $HOME/profile
profile! file $PATH_TO_PLUGIN/*/SimpylFold.vim
```
And upload the file `$HOME/profile`.

**Comment by tmhedberg** [comment]
Ah, the main problem was just a part of my `.vimrc` which was interfering with autoloading the plugin. I removed that, and it now *mostly* works. However, the folding of import blocks is still not working, even if I manually toggle it on and off with the commands you added. Is import folding working for you?

Here's a profile, in case it's useful:

```
SCRIPT  /home/t/dotfiles/.vim/bundle/SimpylFold/autoload/SimpylFold.vim
Sourced 1 time
Total time:   0.000209
 Self time:   0.000209

count  total (s)   self (s)
                            let s:non_blank_regex = '^\s*[^[:space:]#]'
    1              0.000003 let s:multiline_def_end_regex = '):$'
    1              0.000003 let s:multiline_def_end_solo_regex = '^\s*):$'
    1              0.000003 let s:docstring_start_regex = '^\s*[rR]\?\("""\|''''''\)\%(.*\1\s*$\)\@!'
    1              0.000003 let s:docstring_end_single_regex = '''''''\s*$'
    1              0.000003 let s:docstring_end_double_regex = '"""\s*$'
    1              0.000003 let s:import_start_regex = '^\%(from\|import\)'
    1              0.000002 let s:import_cont_regex = 'from.*\((\)[^)]*$\|.*\(\\\)$'
    1              0.000003 let s:import_end_paren_regex = ')\s*$'
    1              0.000003 let s:import_end_esc_regex = '[^\\]$'

                            " Initialize buffer
    1              0.000003 function! SimpylFold#BufferInit() abort
                                if &filetype ==# 'pyrex' || &filetype ==# 'cython'
                                    let b:SimpylFold_def_regex = '\v^\s*%(%(class|%(async\s+)?def|cdef|cpdef|ctypedef)\s+\w+)|cdef\s*:'
                                else
                                    let b:SimpylFold_def_regex = '\v^\s*%(class|%(async\s+)?def)\s+\w+|if\s+__name__\s*\=\=\s*%("__main__"|''__main__'')\s*:'
                                endif

                                if !exists('b:SimpylFold_fold_docstring')
                                    let b:SimpylFold_fold_docstring = !exists('g:SimpylFold_fold_docstring') || g:SimpylFold_fold_docstring
                                endif
                                if !exists('b:SimpylFold_fold_import')
                                    let b:SimpylFold_fold_import = !exists('g:SimpylFold_fold_import') || g:SimpylFold_fold_import
                                endif
                            endfunction

                            " Calculate indent
    1              0.000003 function! s:indent(line) abort
                                let ind = matchend(a:line, '^ *') / &tabstop
                                " Fix indent for solo def multiline endings
                                if a:line =~# s:multiline_def_end_solo_regex
                                    return ind + 1
                                endif
                                return ind
                            endfunction

                            " Determine the number of containing class or function definitions for the
                            " given line.
                            " This function requires that `lnum` is >= previous `lnum`s.
    1              0.000003 function! s:defs(cache, lines, non_blanks, lnum) abort
                                if has_key(a:cache[a:lnum], 'defs')
                                    return a:cache[a:lnum]['defs']
                                endif

                                " Indent level
                                let ind = s:indent(a:lines[a:lnum])
                                let a:cache[a:lnum]['indent'] = ind  " Cache for use in the loop
                                if ind == 0
                                    let a:cache[a:lnum]['defs'] = 0
                                    return 0
                                endif

                                " Walk backwards to find the previous non-blank line with
                                " a lower indent level than this line
                                let non_blanks_prev = a:non_blanks[:index(a:non_blanks, a:lnum) - 1]
                                for lnum_prev in reverse(copy(non_blanks_prev))
                                    if a:cache[lnum_prev]['indent'] < ind
                                        let defs = s:defs(a:cache, a:lines, non_blanks_prev, lnum_prev) +
                                            \ a:cache[lnum_prev]['is_def']
                                        let a:cache[a:lnum]['defs'] = defs
                                        return defs
                                    endif
                                endfor

                                let a:cache[a:lnum]['defs'] = 0
                                return 0
                            endfunction

                            " Construct a foldexpr value and cache it
    1              0.000003 function! s:foldexpr(cache_lnum, foldlevel, is_beginning) abort
                                if a:is_beginning
                                    let a:cache_lnum['foldexpr'] = '>' . a:foldlevel
                                else
                                    let a:cache_lnum['foldexpr'] = a:foldlevel
                                endif
                            endfunction

                            " Create a new cache
    1              0.000002 function! s:cache() abort
                                let cache = [{}]  " With padding for lnum offset
                                let lines = getbufline(bufnr('%'), 1, '$')
                                let lnum_last = len(lines)
                                call insert(lines, '')  " Padding for lnum offset

                                " Cache everything generic that needs to be used later
                                let non_blanks = []
                                for lnum in range(1, lnum_last)
                                    let line = lines[lnum]
                                    if line =~# s:non_blank_regex
                                        call add(non_blanks, lnum)
                                        call add(cache, {'blank': 0, 'is_def': line =~# b:SimpylFold_def_regex})
                                    else
                                        call add(cache, {'blank': 1, 'is_def': 0, 'foldexpr': -1})
                                    endif
                                endfor

                                " Cache non-blanks
                                let in_docstring = 0
                                let in_import = 0
                                for lnum in non_blanks
                                    let line = lines[lnum]

                                    " Docstrings
                                    if b:SimpylFold_fold_docstring
                                        if in_docstring
                                            if line =~# docstring_end_regex
                                                let in_docstring = 0
                                            endif

                                            call s:foldexpr(cache[lnum], s:defs(cache, lines, non_blanks, lnum) + 1, 0)
                                            continue
                                        else
                                            let lnum_prev = lnum - 1
                                            if !cache[lnum_prev]['blank']
                                                let docstring_match = matchlist(line, s:docstring_start_regex)
                                                if !empty(docstring_match) &&
                                                        \ (cache[lnum_prev]['is_def'] ||
                                                        \  lines[lnum_prev] =~# s:multiline_def_end_regex)
                                                    let in_docstring = 1

                                                    if docstring_match[1] ==# '"""'
                                                        let docstring_end_regex = s:docstring_end_double_regex
                                                    else
                                                        let docstring_end_regex = s:docstring_end_single_regex
                                                    endif

                                                    call s:foldexpr(cache[lnum], s:defs(cache, lines, non_blanks, lnum) + 1, 1)
                                                    continue
                                                endif
                                            endif
                                        endif
                                    endif

                                    " Imports
                                    if b:SimpylFold_fold_import
                                        if in_import
                                            if line =~# import_end_regex
                                                let in_import = 0
                                            endif

                                            call s:foldexpr(cache[lnum], s:defs(cache, lines, non_blanks, lnum) + 1, 0)
                                            continue
                                        elseif match(line, s:import_start_regex) != -1
                                            let import_cont_match = matchlist(line, s:import_cont_regex)
                                            if !empty(import_cont_match)
                                                if import_cont_match[1] ==# '('
                                                    let import_end_regex = s:import_end_paren_regex
                                                    let in_import = 1
                                                elseif import_cont_match[2] ==# '\'
                                                    let import_end_regex = s:import_end_esc_regex
                                                    let in_import = 1
                                                endif
                                            endif

                                            if in_import
                                                call s:foldexpr(cache[lnum], s:defs(cache, lines, non_blanks, lnum) + 1, 1)
                                                continue
                                            endif
                                        endif
                                    endif

                                    " Otherwise, its fold level is equal to its number of containing
                                    " definitions, plus 1, if this line starts a definition of its own
                                    call s:foldexpr(
                                        \ cache[lnum],
                                        \ s:defs(cache, lines, non_blanks, lnum) + cache[lnum]['is_def'],
                                        \ cache[lnum]['is_def'],
                                    \ )
                                endfor

                                return cache
                            endfunction

                            " Compute foldexpr for Python code
    1              0.000004 function! SimpylFold#FoldExpr(lnum) abort
                                if !exists('b:SimpylFold_cache')
                                    let b:SimpylFold_cache = s:cache()
                                endif
                                return b:SimpylFold_cache[a:lnum]['foldexpr']
                            endfunction

                            " Recache the buffer
    1              0.000003 function! SimpylFold#Recache() abort
                                if exists('b:SimpylFold_cache')
                                    unlet b:SimpylFold_cache
                                endif
                            endfunction

                            " Compute foldtext by obtaining the first line of the docstring for
                            " the folded class or function, if any exists
    1              0.000002 function! SimpylFold#FoldText() abort
                                let next = nextnonblank(v:foldstart + 1)
                                let docstring = getline(next)
                                let ds_prefix = '^\s*\%(\%(["'']\)\{3}\|[''"]\ze[^''"]\)'
                                if docstring =~# ds_prefix
                                    let quote_char = docstring[match(docstring, '["'']')]
                                    let docstring = substitute(docstring, ds_prefix, '', '')
                                    if docstring !~# s:non_blank_regex
                                        let docstring =
                                            \ substitute(getline(nextnonblank(next + 1)), '^\s*', '', '')
                                    endif
                                    let docstring = substitute(docstring, quote_char . '\{,3}$', '', '')
                                    return ' ' . docstring
                                endif
                                return ''
                            endfunction

FUNCTION  <SNR>109_cache()
Called 1 time
Total time:   1.133077
 Self time:   0.529730

count  total (s)   self (s)
    1              0.000007     let cache = [{}]  " With padding for lnum offset
    1              0.001293     let lines = getbufline(bufnr('%'), 1, '$')
    1              0.000013     let lnum_last = len(lines)
    1              0.000009     call insert(lines, '')  " Padding for lnum offset

                                " Cache everything generic that needs to be used later
    1              0.000005     let non_blanks = []
 3785              0.005865     for lnum in range(1, lnum_last)
 3784              0.011918         let line = lines[lnum]
 3784              0.018632         if line =~# s:non_blank_regex
 2705              0.007965             call add(non_blanks, lnum)
 2705              0.064927             call add(cache, {'blank': 0, 'is_def': line =~# b:SimpylFold_def_regex})
 2705              0.002887         else
 1079              0.004685             call add(cache, {'blank': 1, 'is_def': 0, 'foldexpr': -1})
 1079              0.000990         endif
 3784              0.005548     endfor

                                " Cache non-blanks
    1              0.000002     let in_docstring = 0
    1              0.000002     let in_import = 0
 2706              0.004359     for lnum in non_blanks
 2705              0.009208         let line = lines[lnum]

                                    " Docstrings
 2705              0.005487         if b:SimpylFold_fold_docstring
 2705              0.003083             if in_docstring
   12              0.000058                 if line =~# docstring_end_regex
    4              0.000008                     let in_docstring = 0
    4              0.000004                 endif

   12   0.002364   0.000291                 call s:foldexpr(cache[lnum], s:defs(cache, lines, non_blanks, lnum) + 1, 0)
   12              0.000021                 continue
                                        else
 2693              0.006883                 let lnum_prev = lnum - 1
 2693              0.007210                 if !cache[lnum_prev]['blank']
 1906              0.030785                     let docstring_match = matchlist(line, s:docstring_start_regex)
 1906              0.009616                     if !empty(docstring_match) && (cache[lnum_prev]['is_def'] ||  lines[lnum_prev] =~# s:multiline_def_end_regex)
    4              0.000008                         let in_docstring = 1

    4              0.000009                         if docstring_match[1] ==# '"""'
    4              0.000013                             let docstring_end_regex = s:docstring_end_double_regex
    4              0.000004                         else
                                                        let docstring_end_regex = s:docstring_end_single_regex
                                                    endif

    4   0.000822   0.000132                         call s:foldexpr(cache[lnum], s:defs(cache, lines, non_blanks, lnum) + 1, 1)
    4              0.000009                         continue
                                                endif
 1902              0.001444                 endif
 2689              0.002068             endif
 2689              0.001963         endif

                                    " Imports
 2689              0.004670         if b:SimpylFold_fold_import
 2689              0.002897             if in_import
                                            if line =~# import_end_regex
                                                let in_import = 0
                                            endif

                                            call s:foldexpr(cache[lnum], s:defs(cache, lines, non_blanks, lnum) + 1, 0)
                                            continue
                                        elseif match(line, s:import_start_regex) != -1
    5              0.000098                 let import_cont_match = matchlist(line, s:import_cont_regex)
    5              0.000013                 if !empty(import_cont_match)
                                                if import_cont_match[1] ==# '('
                                                    let import_end_regex = s:import_end_paren_regex
                                                    let in_import = 1
                                                elseif import_cont_match[2] ==# '\'
                                                    let import_end_regex = s:import_end_esc_regex
                                                    let in_import = 1
                                                endif
                                            endif

    5              0.000005                 if in_import
                                                call s:foldexpr(cache[lnum], s:defs(cache, lines, non_blanks, lnum) + 1, 1)
                                                continue
                                            endif
    5              0.000003             endif
 2689              0.002250         endif

                                    " Otherwise, its fold level is equal to its number of containing
                                    " definitions, plus 1, if this line starts a definition of its own
 2689   0.710844   0.110260         call s:foldexpr( cache[lnum], s:defs(cache, lines, non_blanks, lnum) + cache[lnum]['is_def'], cache[lnum]['is_def'], )
 2689              0.002632     endfor

    1              0.000001     return cache

FUNCTION  <SNR>109_defs()
Called 4590 times
Total time:   0.591010
 Self time:   0.562321

count  total (s)   self (s)
 4590              0.020132     if has_key(a:cache[a:lnum], 'defs')
 1885              0.005501         return a:cache[a:lnum]['defs']
                                endif

                                " Indent level
 2705   0.059858   0.015485     let ind = s:indent(a:lines[a:lnum])
 2705              0.011448     let a:cache[a:lnum]['indent'] = ind  " Cache for use in the loop
 2705              0.003864     if ind == 0
  820              0.002726         let a:cache[a:lnum]['defs'] = 0
  820              0.000977         return 0
                                endif

                                " Walk backwards to find the previous non-blank line with
                                " a lower indent level than this line
 1885              0.141215     let non_blanks_prev = a:non_blanks[:index(a:non_blanks, a:lnum) - 1]
13387              0.108395     for lnum_prev in reverse(copy(non_blanks_prev))
13387              0.040003         if a:cache[lnum_prev]['indent'] < ind
 1885              0.011101             let defs = s:defs(a:cache, a:lines, non_blanks_prev, lnum_prev) + a:cache[lnum_prev]['is_def']
 1885              0.007597             let a:cache[a:lnum]['defs'] = defs
 1885              0.002623             return defs
                                    endif
11502              0.009950     endfor

                                let a:cache[a:lnum]['defs'] = 0
                                return 0

FUNCTION  <SNR>109_foldexpr()
Called 2705 times
Total time:   0.027840
 Self time:   0.027840

count  total (s)   self (s)
 2705              0.006387     if a:is_beginning
   88              0.000355         let a:cache_lnum['foldexpr'] = '>' . a:foldlevel
   88              0.000094     else
 2617              0.007906         let a:cache_lnum['foldexpr'] = a:foldlevel
 2617              0.002348     endif

FUNCTION  SimpylFold#FoldText()
Called 4 times
Total time:   0.000274
 Self time:   0.000274

count  total (s)   self (s)
    4              0.000033     let next = nextnonblank(v:foldstart + 1)
    4              0.000013     let docstring = getline(next)
    4              0.000009     let ds_prefix = '^\s*\%(\%(["'']\)\{3}\|[''"]\ze[^''"]\)'
    4              0.000053     if docstring =~# ds_prefix
    3              0.000025         let quote_char = docstring[match(docstring, '["'']')]
    3              0.000030         let docstring = substitute(docstring, ds_prefix, '', '')
    3              0.000016         if docstring !~# s:non_blank_regex
                                        let docstring = substitute(getline(nextnonblank(next + 1)), '^\s*', '', '')
                                    endif
    3              0.000042         let docstring = substitute(docstring, quote_char . '\{,3}$', '', '')
    3              0.000006         return ' ' . docstring
                                endif
    1              0.000001     return ''

FUNCTION  SimpylFold#Recache()
Called 0 times
Total time:   0.000000
 Self time:   0.000000

count  total (s)   self (s)
                                if exists('b:SimpylFold_cache')
                                    unlet b:SimpylFold_cache
                                endif

FUNCTION  SimpylFold#FoldExpr()
Called 3889 times
Total time:   1.172089
 Self time:   0.039012

count  total (s)   self (s)
 3889              0.011528     if !exists('b:SimpylFold_cache')
    1   1.133346   0.000269         let b:SimpylFold_cache = s:cache()
    1              0.000002     endif
 3889              0.012704     return b:SimpylFold_cache[a:lnum]['foldexpr']

FUNCTION  SimpylFold#BufferInit()
Called 1 time
Total time:   0.000033
 Self time:   0.000033

count  total (s)   self (s)
    1              0.000004     if &filetype ==# 'pyrex' || &filetype ==# 'cython'
                                    let b:SimpylFold_def_regex = '\v^\s*%(%(class|%(async\s+)?def|cdef|cpdef|ctypedef)\s+\w+)|cdef\s*:'
                                else
    1              0.000004         let b:SimpylFold_def_regex = '\v^\s*%(class|%(async\s+)?def)\s+\w+|if\s+__name__\s*\=\=\s*%("__main__"|''__main__'')\s*:'
    1              0.000000     endif

    1              0.000004     if !exists('b:SimpylFold_fold_docstring')
    1              0.000005         let b:SimpylFold_fold_docstring = !exists('g:SimpylFold_fold_docstring') || g:SimpylFold_fold_docstring
    1              0.000001     endif
    1              0.000002     if !exists('b:SimpylFold_fold_import')
    1              0.000005         let b:SimpylFold_fold_import = !exists('g:SimpylFold_fold_import') || g:SimpylFold_fold_import
    1              0.000001     endif

FUNCTION  <SNR>109_indent()
Called 2705 times
Total time:   0.044373
 Self time:   0.044373

count  total (s)   self (s)
 2705              0.017632     let ind = matchend(a:line, '^ *') / &tabstop
                                " Fix indent for solo def multiline endings
 2705              0.013652     if a:line =~# s:multiline_def_end_solo_regex
                                    return ind + 1
                                endif
 2705              0.003594     return ind

FUNCTIONS SORTED ON TOTAL TIME
count  total (s)   self (s)  function
 3889   1.172089   0.039012  SimpylFold#FoldExpr()
    1   1.133077   0.529730  <SNR>109_cache()
 4590   0.591010   0.562321  <SNR>109_defs()
 2705   0.044373             <SNR>109_indent()
 2705   0.027840             <SNR>109_foldexpr()
    4   0.000274             SimpylFold#FoldText()
    1   0.000033             SimpylFold#BufferInit()
                             SimpylFold#Recache()

FUNCTIONS SORTED ON SELF TIME
count  total (s)   self (s)  function
 4590   0.591010   0.562321  <SNR>109_defs()
    1   1.133077   0.529730  <SNR>109_cache()
 2705              0.044373  <SNR>109_indent()
 3889   1.172089   0.039012  SimpylFold#FoldExpr()
 2705              0.027840  <SNR>109_foldexpr()
    4              0.000274  SimpylFold#FoldText()
    1              0.000033  SimpylFold#BufferInit()
                             SimpylFold#Recache()
```

**Comment by nfnty** [comment]
Alright. It seems I've changed the import logic to only fold multiline imports. Should we have 2 different import fold methods? I prefer it like this.

PS. Next time upload a gist instead to not clutter the thread.

**Comment by tmhedberg** [comment]
The import block folding exists in its current form because several users requested it. It's not something I personally have strong feelings about, but if it stops working, some people will be unhappy. Please restore the existing behavior, and if you want to add an optional alternative way of folding imports, then do that in a separate pull request.

**Comment by nfnty** [comment]
Ok, fixed.

**Comment by tmhedberg** [comment]
Not sure why I didn't notice this yesterday, but class folding appears broken as well. Specifically, the plugin seems to create a single-line fold for each class which only encompasses the first line of the class (e.g. `class Foo(object):`). The rest of the class is not contained in the fold.

Additionally, in some cases, there are single-line folds being created for the blank lines that separate method definitions. Once again, I'm using [`categorical.py`](https://raw.githubusercontent.com/mwaskom/seaborn/master/seaborn/categorical.py) as a test case. In that file, for instance, I see a single line fold on the blank line between the method definitions for `CategoricalPlotter.establish_variables` and `_CategoricalPlotter._group_longform` (among others).

You may have to enable `foldcolumn` or set `foldminlines` to 0 in order to notice these spurious folds.

**Comment by nfnty** [comment]
Can't reproduce the class issue.

I'll fix the blank single line folding issue.

**Comment by nfnty** [comment]
Blank single line folding issue fixed.

File: autoload/SimpylFold.vim [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+        let b:SimpylFold_fold_docstring = !exists('g:SimpylFold_fold_docstring') || g:SimpylFold_fold_docstring
+    endif
+    if !exists('b:SimpylFold_fold_import')
+        let b:SimpylFold_fold_import = !exists('g:SimpylFold_fold_import') || g:SimpylFold_fold_import
+    endif
+endfunction
+
+" Calculate indent
+function! s:indent(line) abort
+    let ind = matchend(a:line, '^ *') / &tabstop
```
Comment (tmhedberg) [inline autoload/SimpylFold.vim]: I think the problem I mentioned with class folding is due to this line. I have `tabstop` set to 8, but use 4-space indents. I think you should divide by the value of `softtabstop` anyway. `tabstop` is the "number of spaces that a <Tab> in the file counts for", according to the documentation. That means a literal tab character, not the spaces inserted by pressing the tab key with `expandtab` enabled, which I imagine is usually the case when editing Python.

Comment (nfnty) [inline autoload/SimpylFold.vim]: Ah, yes of course.

Comment (nfnty) [inline autoload/SimpylFold.vim]: Fixed.

**Comment by tmhedberg** [comment]
I can't find any more issues. Looks good now. Thanks a lot!

**Comment by nfnty** [comment]
Nice! I'm working on refactoring `s:defs()` to become non-recursive.

**Pull request merged**